### DISCUSSION TITLE
More frequent pre-releases

### DISCUSSION THREAD

**Discussion by casey** [rust-bitcoin/rust-bitcoin#1520] [Announcements]
This conversation started in #1377, but I wanted to give it its own issue. `rust-bitcoin` currently has a relatively slow release schedule, for good reason, to avoid churn while making large breaking changes. However, this can mean that dependents that need recently added features (e.g. me) have to maintain their own forks of `rust-bitcoin` and publish them to crates.io in order to publish dependent crates, since crates.io does not allow non-crates.io dependencies. (Otherwise we could just use a git-dependency pointing to a commit on GitHub.)

I wanted to propose more frequent pre-releases, perhaps on a monthly or six-week cadence. If these pre-releases use a pre-release suffix, e.g. `0.30.0-pre1`, then existing projects using semver won't pick them up, but a project that is willing to deal with breaking changes can specify the pre-release version explicitly.

Maintainers could publish current master without worrying too much about whether it was a breaking or non-breaking change, and fast moving dependents could hopefully avoid needing to maintaining and publishing their own forks.

This would be an additional burden on the maintainers, and if it's just me that would avoid maintaining my own fork by doing this, it might not be worth it.

**Comment by apoelstra** [comment]
I like this idea. I wouldn't mind having a bunch of `-preX` versions out there even if they were in broken/inconsistent states.

**Comment by tcharding** [comment]
I like it, I'm happy to do any additional work required for this.

**Comment by apoelstra** [comment]
I think we were planning to do a six-month release cadence -- so we could release a prerelease version on the first of every month (or whatever), with the understanding that `pre6` was going to be the real release so we need to define a feature set and close out any problems.

Or we could just do a prerelease whenever somebody asked for one.

**Comment by Kixunil** [comment]
I think the only thing I'm unsure about is the changelog. Releasing without changelog sucks but so does making the changelog each time and then probably doing union of all pre-releases in the final release.

> Or we could just do a prerelease whenever somebody asked for one.

That sounds like a reasonable compromise.

**Reply by apoelstra** [reply]
I'd be fine releasing without a changelog. The -pre releases are understood to exist for people who are trying to get a jump on things (and/or people trying to update their rust-bitcoin dependency across multiple commits).

As long as we don't get into a habit of leaving -pre releases open for years and encouraging people to use them (*cough* clap, rocket) this shouldn't be a problem.

**Reply by Kixunil** [reply]
I'd like to have at least a few important changes written somewhere for context but I guess we don't have to be super strict.

**Comment by tcharding** [comment]
Whats the verdict on pre-releases, every month or when someone asks? If the latter, how are devs supposed to know they can ask for a pre-release?

**Reply by Kixunil** [reply]
IMO every month is frequent and burdensome for us. Just look how long it took us to release 0.30 from the moment we decided to push for it. We can document it somewhere.

**Reply by apoelstra** [reply]
I do think we should try to do this though, because it's valuable to practice releases.

Want to aim for every 2 months? So we'd be going for:

* 0.30.0 (March 1, lol)
* 0.31.0pre1 (May 1)
* 0.31.0pre2 (July 1) (feature freeze after this one)
* 0.31.0 (Sept 1)

This gives us a 2-month window in which we can slip release dates, but still gives us reasonably frequent practice at doing releases. The last release was only delayed a week or so because we wanted extra stuff to get in. Most of the delay was due to the `bitcoin-internals` thing then us getting out of sync because we'd delayed into peoples' vacation time and were trying to sync up the blog post and deal with crates.io things, and releasing multiple crates without remembering what order they all had to go in.

IOW with practice, and if we don't get crates.io-sniped again, we should be able to go much faste.

**Reply by Kixunil** [reply]
2 months sounds better, maybe even three would be OK.

**Reply by apoelstra** [reply]
For this cycle let's do a pre-release in 2 months (or 5 weeks, rather -- on May 1st), a real release in 4 months, and revisit after that (see other convo, in which I mentioned that I will have a newborn around Sep 1 so we can't do a real release on that data). How does that sound?

**Reply by Kixunil** [reply]
Sounds good!

**Reply by tcharding** [reply]
Lets go!

**Reply by tcharding** [reply]
Does that mean `rust-bitcoin` out on May 1st? If so we will likely need to start a week earlier on secp and hashes (and hopefully hex) if they need releasing also. Or does May 1st mean try to release the lowest crate on May 1st?

**Reply by apoelstra** [reply]
Let's say, "release the lowest crate on May 1st". We should strive to do it all in one day, which I think we can do if we correctly figure out what needs to be released in what order.

**Reply by tcharding** [reply]
Cool

**Comment by tcharding** [comment]
We should probably start work on the first pre-release soonish eh? Can we aim to start at the bottom of the stack by end of the month? `rust-bech32` and `hex-conservative` both have interesting changes on `master`.

**Reply by apoelstra** [reply]
Yep, was just thinking that we should start pushing on that today.

**Comment by tcharding** [comment]
We failed at this. Please see #5647.
